python/mpc_cvxgen/utils.py

```diff
@@ -0,0 +1,392 @@
+        """
+        if t < self.x[0]:
+            return None
+        elif t > self.x[-1]:
+            return None
+        i = self._search_index(t)
+        dx = t - self.x[i]
+        der0 = \
+            self.a[i] + self.b[i] * dx + \
+            self.c[i] * dx ** 2.0 + self.d[i] * dx ** 3.0
+        return der0
+
+    def calc_der1(self, t):
+        """ Calculate the 1st derivative evaluated at t.
+
+        Args:
+            t (:float:): position along the 1-d spline
+
+        Returns:
+            der1 (:float:): 1st derivative evaluated at t
+        """
+        if t < self.x[0]:
+            return None
+        elif t > self.x[-1]:
+            return None
+        i = self._search_index(t)
+        dx = t - self.x[i]
+        der1 = \
+            self.b[i] + 2.0 * self.c[i] * dx + \
+            3.0 * self.d[i] * dx ** 2.0
+        return der1
+
+    def calc_der2(self, t):
+        """ Calculate the 2nd derivative evaluated at t.
+
+        Args:
+            t (:float:): position along the 1-d spline
+
+        Returns:
+            der2 (:float:): 2nd derivative evaluated at t
+        """
+        if t < self.x[0]:
+            return None
+        elif t > self.x[-1]:
+            return None
+        i = self._search_index(t)
+        dx = t - self.x[i]
+        der2 = 2.0 * self.c[i] + 6.0 * self.d[i] * dx
+        return der2
+
+    def _search_index(self, x):
+        """ Search the spline for index closest to x.
+
+        Args:
+            x (:float:): position along the 1-d spline
+
+        Returns:
+            (:int:): index closest to x
+        """
+        return bisect.bisect(self.x, x) - 1
+
+    def _matrix_a(self, h):
+        """Create the constants matrix a used in spline construction.
+
+        Args:
+            h (:np.ndarray:): list of deltas between values
+
+        Returns:
+            matrix_a (:np.ndarray:): constants matrix
+        """
+        matrix_a = np.zeros((self.nx, self.nx))
+        matrix_a[0, 0] = 1.0
+        for i in range(self.nx - 1):
+            if i != (self.nx - 2):
+                matrix_a[i + 1, i + 1] = 2.0 * (h[i] + h[i + 1])
+            matrix_a[i + 1, i] = h[i]
+            matrix_a[i, i + 1] = h[i]
+
+        matrix_a[0, 1] = 0.0
+        matrix_a[self.nx - 1, self.nx - 2] = 0.0
+        matrix_a[self.nx - 1, self.nx - 1] = 1.0
+        return matrix_a
+
+    def _matrix_b(self, h):
+        """ Create the 1st derivative matrix b used in spline construction.
+
+        Args:
+            h (:np.ndarray:): list of deltas between values
+
+        Returns:
+            matrix_b (:np.ndarray:): 1st derivative matrix
+        """
+        matrix_b = np.zeros(self.nx)
+        for i in range(self.nx - 2):
+            matrix_b[i + 1] = \
+                3.0 * (self.a[i + 2] - self.a[i + 1]) / \
+                h[i + 1] - 3.0 * (self.a[i + 1] - self.a[i]) / h[i]
+        return matrix_b
+
+
+class CubicSpline2D:
```

Comment:
Not sure if you've encountered any issues with this implementation, but my [C++ implementation](https://github.com/erdos-project/frenet_optimal_trajectory_planner/tree/master/src/CubicSpline) faced issues when there were lots of collinear points (ie. straight line). I had to remove them to improve the stability of the spline. 

I would try this out by just making a very long line with points spaced 0.5-1m apart and see if the cubic spline interpolation is reasonable.